File: rpipico/lib/mq135.py
"""
MQ135 Air Quality Sensor Library
Library for reading air quality data from MQ135 sensor via ADC
"""

from machine import Pin, ADC

# Float comparison epsilon
EPSILON = 1e-6

# ADC constants
ADC_MAX_VALUE = 65535
VOLTAGE_REFERENCE = 3.3

class MQ135:
    def __init__(self, adc_pin, r_zero=42304.5, r_load=10000):
        """
        Initialize MQ135 sensor
        
        Args:
            adc_pin: GPIO pin number for ADC input
            r_zero: Calibrated resistance in clean air (default from config)
            r_load: Load resistor value in ohms (default 10kΩ)
        """
        self.adc = ADC(Pin(adc_pin))
        self.r_load = r_load
        self.r_zero = r_zero
        
    def read_voltage(self):
        """Read voltage from ADC"""
        raw = self.adc.read_u16()
        voltage = (raw / ADC_MAX_VALUE) * VOLTAGE_REFERENCE
        return voltage, raw
    
    def read_resistance(self):
        """Calculate sensor resistance"""
        voltage, raw = self.read_voltage()
        # Fix float comparison: use epsilon instead of <= 0.01
        if voltage < 0.01 + EPSILON:
            return float('inf'), voltage, raw
        resistance = ((VOLTAGE_REFERENCE - voltage) / voltage) * self.r_load
        return resistance, voltage, raw
    
    def read_ratio(self):
        """Read Rs/R0 ratio"""
        resistance, voltage, raw = self.read_resistance()
        # Fix float comparison: check for infinity and zero with epsilon
        if resistance == float('inf') or abs(self.r_zero) < EPSILON:
            return 0, resistance, voltage, raw
        ratio = resistance / self.r_zero
        return ratio, resistance, voltage, raw
    
    def read_co2_ppm(self):
        """Estimate CO2 concentration in ppm"""
        ratio, resistance, voltage, raw = self.read_ratio()
        return self._calculate_co2_ppm(ratio), ratio, resistance, voltage, raw
    
    def read_nh3_ppm(self):
        """Estimate NH3 concentration in ppm"""
        ratio, resistance, voltage, raw = self.read_ratio()
        return self._calculate_nh3_ppm(ratio), ratio, resistance, voltage, raw
    
    def read_alcohol_ppm(self):
        """Estimate alcohol concentration in ppm"""
        ratio, resistance, voltage, raw = self.read_ratio()
        return self._calculate_alcohol_ppm(ratio), ratio, resistance, voltage, raw
    
    def get_air_quality_status(self, co2_ppm):
        """
        Get air quality status based on CO2 levels
        
        Returns:
            tuple: (status_string, air_quality_index)
        """
        if co2_ppm < 400:
            return 'Excellent', 1
        elif co2_ppm < 600:
            return 'Good', 2
        elif co2_ppm < 1000:
            return 'Fair', 3
        elif co2_ppm < 1500:
            return 'Poor', 4
        elif co2_ppm < 2500:
            return 'Very Poor', 5
        else:
            return 'Hazardous', 6
    
    def get_all_readings(self):
        """Get all sensor readings as a dictionary"""
        ratio, resistance, voltage, raw = self.read_ratio()
        
        co2_ppm = self._calculate_co2_ppm(ratio)
        nh3_ppm = self._calculate_nh3_ppm(ratio)
        alcohol_ppm = self._calculate_alcohol_ppm(ratio)
        status, aqi = self.get_air_quality_status(co2_ppm)
        
        return {
            'raw_adc': raw,
            'voltage_v': round(voltage, 3),
            'resistance_ohm': round(resistance, 1),
            'ratio_rs_r0': round(ratio, 3),
            'co2_ppm': round(co2_ppm, 1),
            'nh3_ppm': round(nh3_ppm, 1),
            'alcohol_ppm': round(alcohol_ppm, 1),
            'air_quality_status': status,
            'air_quality_index': aqi,
            'r_zero_ohm': self.r_zero
        }
    
    def _calculate_co2_ppm(self, ratio):
        """Calculate CO2 ppm from ratio"""
        if ratio < EPSILON:
            return 0
        ppm = 116.6020682 * (ratio ** -2.769034857)
        return max(0, min(10000, ppm))
    
    def _calculate_nh3_ppm(self, ratio):
        """Calculate NH3 ppm from ratio"""
        if ratio < EPSILON:
            return 0
        ppm = 102.694 * (ratio ** -2.815)
        return max(0, min(500, ppm))
    
    def _calculate_alcohol_ppm(self, ratio):
        """Calculate alcohol ppm from ratio"""
        if ratio < EPSILON:
            return 0
        ppm = 77.255 * (ratio ** -3.18)
        return max(0, min(1000, ppm))

File: airsensors/ardudueair/MQ135.cpp
#include "MQ135.h"
#include "Config.h"

MQ135::MQ135(int adcPin, float rZero, float rLoad)
    : adcPin(adcPin), r_load(rLoad), r_zero(rZero),
      cal_r_zero(rZero), cal_temp_c(20.0f), cal_humidity(50.0f), calibrated(false) {
  pinMode(adcPin, INPUT);
}

float MQ135::read_voltage() {
  int raw = analogRead(adcPin);
  float voltage = (raw / (float)ADC_MAX_VALUE) * VOLTAGE_REFERENCE;
  return voltage;
}

float MQ135::read_resistance() {
  float voltage = read_voltage();
  if (voltage < MIN_VOLTAGE_THRESHOLD + EPSILON) {
    return INFINITY;
  }
  float resistance = ((VOLTAGE_REFERENCE - voltage) / voltage) * r_load;
  return resistance;
}

float MQ135::read_ratio() {
  float resistance = read_resistance();
  if (isinf(resistance) || fabs(r_zero) < EPSILON) {
    return 0;
  }
  return resistance / r_zero;
}

float MQ135::read_co2_ppm() {
  float ratio = read_ratio();
  return _calculate_co2_ppm(ratio);
}

float MQ135::read_nh3_ppm() {
  float ratio = read_ratio();
  return _calculate_nh3_ppm(ratio);
}

float MQ135::read_alcohol_ppm() {
  float ratio = read_ratio();
  return _calculate_alcohol_ppm(ratio);
}

const char* MQ135::get_air_quality_status(float co2_ppm) const {
  if (co2_ppm < AQ_EXCELLENT) return "Excellent";
  if (co2_ppm < AQ_GOOD) return "Good";
  if (co2_ppm < AQ_FAIR) return "Fair";
  if (co2_ppm < AQ_POOR) return "Poor";
  if (co2_ppm < AQ_VERY_POOR) return "Very Poor";
  return "Hazardous";
}

int MQ135::get_air_quality_index(float co2_ppm) const {
  if (co2_ppm < AQ_EXCELLENT) return 1;
  if (co2_ppm < AQ_GOOD) return 2;
  if (co2_ppm < AQ_FAIR) return 3;
  if (co2_ppm < AQ_POOR) return 4;
  if (co2_ppm < AQ_VERY_POOR) return 5;
  return 6;
}

float MQ135::_calculate_co2_ppm(float ratio) const {
  if (ratio < EPSILON) return 0;
  float ppm = MQ135_CO2_A * pow(ratio, MQ135_CO2_B);
  return max(0.0f, min(MQ135_CO2_MAX, ppm));
}

float MQ135::_calculate_nh3_ppm(float ratio) const {
  if (ratio < EPSILON) return 0;
  float ppm = MQ135_NH3_A * pow(ratio, MQ135_NH3_B);
  return max(0.0f, min(MQ135_NH3_MAX, ppm));
}

float MQ135::_calculate_alcohol_ppm(float ratio) const {
  if (ratio < EPSILON) return 0;
  float ppm = MQ135_ALCOHOL_A * pow(ratio, MQ135_ALCOHOL_B);
  return max(0.0f, min(MQ135_ALCOHOL_MAX, ppm));
}

float MQ135::_compensate_ratio(float ratio, float temp_c, float humidity_percent) const {
  if (!calibrated) return ratio;
  // Empirical T/RH compensation relative to calibration baseline.
  // Resistance drifts ~2%/°C and ~1.5%/%RH from baseline.
  float temp_factor = 1.0f - 0.02f * (temp_c - cal_temp_c);
  float hum_factor  = 1.0f - 0.015f * (humidity_percent - cal_humidity);
  float factor = temp_factor * hum_factor;
  if (fabs(factor) < EPSILON) return ratio;
  return ratio / factor;
}

MQ135Readings MQ135::get_all_readings() {
  MQ135Readings r;
  r.raw_adc = analogRead(adcPin);
  r.voltage_v = (r.raw_adc / (float)ADC_MAX_VALUE) * VOLTAGE_REFERENCE;
  if (r.voltage_v < MIN_VOLTAGE_THRESHOLD + EPSILON) {
    r.resistance_ohm = INFINITY;
    r.ratio_rs_r0 = 0;
  } else {
    r.resistance_ohm = ((VOLTAGE_REFERENCE - r.voltage_v) / r.voltage_v) * r_load;
    r.ratio_rs_r0 = (fabs(r_zero) < EPSILON) ? 0 : r.resistance_ohm / r_zero;
  }
  r.co2_ppm = _calculate_co2_ppm(r.ratio_rs_r0);
  r.nh3_ppm = _calculate_nh3_ppm(r.ratio_rs_r0);
  r.alcohol_ppm = _calculate_alcohol_ppm(r.ratio_rs_r0);
  r.air_quality_status = get_air_quality_status(r.co2_ppm);
  r.air_quality_index = get_air_quality_index(r.co2_ppm);
  r.r_zero_ohm = r_zero;
  return r;
}

MQ135Readings MQ135::get_compensated_readings(float temp_c, float humidity_percent) {
  MQ135Readings r = get_all_readings();
  if (!calibrated) return r;

  float compensated_ratio = _compensate_ratio(r.ratio_rs_r0, temp_c, humidity_percent);
  r.ratio_rs_r0 = compensated_ratio;
  r.co2_ppm = _calculate_co2_ppm(compensated_ratio);
  r.nh3_ppm = _calculate_nh3_ppm(compensated_ratio);
  r.alcohol_ppm = _calculate_alcohol_ppm(compensated_ratio);
  r.air_quality_status = get_air_quality_status(r.co2_ppm);
  r.air_quality_index = get_air_quality_index(r.co2_ppm);
  return r;
}

void MQ135::auto_calibrate(float temp_c, float humidity_percent, float assumed_co2_ppm) {
  if (assumed_co2_ppm < EPSILON || assumed_co2_ppm > MQ135_CO2_MAX) return;

  // Average multiple readings for a stable baseline
  float resistance_sum = 0;
  int valid_samples = 0;
  for (int i = 0; i < MQ135_CALIBRATION_SAMPLES; i++) {
    float r = read_resistance();
    if (!isinf(r) && r >= EPSILON) {
      resistance_sum += r;
      valid_samples++;
    }
    delay(MQ135_CALIBRATION_SAMPLE_DELAY_MS);
  }
  if (valid_samples == 0) return;
  float resistance = resistance_sum / valid_samples;

  // Inverse of: ppm = A * ratio^B  =>  ratio = (ppm / A)^(1/B)
  float target_ratio = pow(assumed_co2_ppm / MQ135_CO2_A, 1.0f / MQ135_CO2_B);
  if (target_ratio < EPSILON) return;

  cal_r_zero = resistance / target_ratio;
  cal_temp_c = temp_c;
  cal_humidity = humidity_percent;
  r_zero = cal_r_zero;
  calibrated = true;
}

bool MQ135::is_calibrated() const {
  return calibrated;
}

bool MQ135::is_present() const {
  // A floating or disconnected pin reads near 0V or rail; a connected MQ135
  // pulls the analog pin into a mid-range voltage. Average 5 reads for stability.
  long sum = 0;
  for (int i = 0; i < 5; i++) {
    sum += analogRead(adcPin);
    delay(10);
  }
  float voltage = (sum / 5.0f / ADC_MAX_VALUE) * VOLTAGE_REFERENCE;
  return voltage > 0.05f && voltage < (VOLTAGE_REFERENCE - 0.05f);
}
